docs/madr/decisions/003-logging.md

```diff
@@ -0,0 +1,220 @@
+log.Info("proxy reconfigured", ...)
+```
+_BAD_
+```go
+log.Info("OnStreamResponse", ...)
+```
+
+**DEBUG**
+DEBUG should print information that can help troubleshoot the issue.
+Here we can log actions that do not modify the system.
+_Examples:_
+* GET request to the API Server
+* We generated the Envoy configuration, but it's the same so nothing significant happens
+* We triggered XDS generation because resources have changed, but it does not mean that proxy will be reconfigured.
+* Reconciliation process is run on interval or event. User should not care until it actually does something important.
+* `else conditions`. For example, "pod already has Kuma sidecar. Skipping the injection"
+* Content of the update. 
+  For example, on the INFO we may log that "user X modified resource Y", but logging the whole resource is heavy.
+  We can log the content of the resource on DEBUG level.
+
+These logs are meant to be read by Kuma maintainers and users that are ok with looking at Kuma implementation to understand them.
+
+**WARNING**
+We don't have a warning level in Kuma.
+Warning could be a level to inform a user about potential misuse of the system like:
+* Using deprecated options
+* Warning about usage of autogenerated self-signed certs.
+
+If you really need this level, prefix INFO level with `[WARNING]`.
+Consider other mechanism of informing a user about it, it's easy to skip the log by accident.
+
+When still in doubt:
+* log on INFO but with separate logger name, so the user may disable this log.
+  _Note: controlling the log level on logger name is not yet available._
+* emit a metric with generalized information, so even if user disable the log, they still have some data.
+
+#### Log or emit a metric
+
+Metrics and logs are closely related, yet they have different characteristics. 
+Logs give you the best precision. You know exactly what happened at specific time.
+Logs usually give you more information than metrics. For example "User X accessed endpoint /xyz".
+Metrics are more generalized so "Someone accessed endpoint /xyz".
+In majority of cases, (structured) logging is superior, but it comes at cost of maintain the infrastructure to collect them.
+
+For example, you may want to log every GET request to your server.
+With systems that take a significant load, you may easily DDoS your logging system.
+With metrics, you can increase counters and let the scraper fetch them every X second.
+
+If you want to introduce a log that can be noisy, consider adding a generalized metric.
+This way even if the user disables this log, they still have some data to troubleshoot. 
+
+#### Global variable or pass the logger
+
+In Kuma, we can either declare logger globally in the package
+```go
+var logger = core.Log.WithName("xds")
+```
+or pass it explicitly to the component
+```go
+NewComponent(core.Log.WithName("xds"))
+```
+
+The library code should ALWAYS receive the logger. The user of the library should decide whether they want to log or not.
```

Comment:
`library code` - what do you mean by that? is that packages in Kuma that could be reused somewhere else?

Reply:
Like I said in the MADR, I'd consider `pkg/core` and `pkg/utils` a library code. However, it is not explicit.
`pkg/core` is used heavily as library code in Kong Mesh which is built on top of Kuma.

We did the mistake of using `pkg` as the "default" directory for code instead of `internal` which seems to be a pattern in more modern Go projects. This way `internal` is internal code and `pkg` is an explicit library code.